docs/user_guide/building_blocks/producers.md

```diff
@@ -0,0 +1,176 @@
+    uses={"Jet.pt"},
+    # declare which columns are created by this Producer
+    produces={"HT"},
+)
+def HT_feature(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    # reconstruct HT and write in the events
+    events = set_ak_column(events, "HT", ak.sum(events.Jet.pt, axis=1))
+
+    return events
+
+
+@producer(
+    # declare which columns are needed for this Producer, if a Producer is given, takes all columns
+    # declared in the corresponding field from the given Producer
+    uses={HT_feature},
+    # declare which columns are created by this Producer, if a Producer is given, takes all columns
+    # declared in the corresponding field from the given Producer
+    produces={HT_feature, "Jet.pt_squared"},
+)
+def all_features(self: Producer, events: ak.Array, **kwargs) -> ak.Array:
+    # use other producer to create HT column
+    events = self[HT_feature](events, **kwargs)
+
+    # create for all jets a column containing the square of the transverse momentum
+    events = set_ak_column(events, "Jet.pt_squared", events.Jet.pt * events.Jet.pt)
+
+    return events
+```
+
+The ```all_features``` producer creates therefore two new columns, the ```HT``` column on event
+level, and the ```pt_squared``` column for each object of the ```Jet``` collection.
+
+Notes:
+- If one wants to index the ```events``` array in a way that the index does not exist for every
+event (e.g. ```events.Jet.pt[10]``` does only work for events with 11 or more jets), it is possible
+to use the {py:class}`~columnflow.columnar_util.Route` class and its
+{py:meth}`~columnflow.columnar_util.Route.apply` function, which allows to give a default value to
+the returned array in the case of a missing element without throwing an error.
+
+- If the {py:class}`~columnflow.production.Producer` is built in a new file and to be used directly
+by {py:class}`~columnflow.tasks.production.ProduceColumns`, you will still need to put the name of
+the new file along with its path in the ```law.cfg``` file under the ```production_modules```
+argument for law to be able to find the file.
+
+- If you want to use some fields, like the ```Jet``` field, as a Lorentz vector to apply operations
+on, you might use the {py:func}`~columnflow.production.util.attach_coffea_behavior` function. This
+function can be applied on the ```events``` array using
+```python
+events = self[attach_coffea_behavior](events, **kwargs)
+```
+If the name of the field does not correspond to a standard field name, e.g. "BtaggedJets", which
+should provide the same behaviour as a normal jet, the behaviour can still be set, using
+```python
+collections = {x: {"type_name": "Jet"} for x in ["BtaggedJets"]}
+events = self[attach_coffea_behavior](events, collections=collections, **kwargs)
+```
+
+- The weights for plotting should ideally be created in the
+{py:class}`~columnflow.tasks.production.ProduceColumns` task when possible, after the selection and
+reduction of the data.
```

Comment:
true, I might just let the one new item in this list here and link to the best practices section for the rest, what do you think? For now I am just changing all comments in the same way

Reply:
done, I let the law.cfg note there since it is slightly different for each object and is needed for the running, not just a convenience tool like the other comments, the rest is mostly put in the best_practice section and only a link to this section is provided in the notes